### ISSUE TITLE
Top-K Sampling Support

### ISSUE THREAD

**Issue by mikemykhaylov** [lmstudio-ai/mlx-engine#59]
Hello,

I find that for some models, like Mistral Nemo, it is very beneficial to restrict the number of considered completions to improve the coherence of the output. However, compared to llama.cpp, MLX does not seem to support top K sampling in LM Studio. It does look like the underlying library [supports that](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.topk.html), so implementing that would be much appreciated.

llama.cpp models support Top-K
<img width="429" alt="Screenshot 2024-12-15 at 12 29 11" src="https://github.com/user-attachments/assets/9f126d00-0604-4ff2-a0d8-006991bc5b71" />

MLX models do not support Top-K
<img width="425" alt="Screenshot 2024-12-15 at 12 29 44" src="https://github.com/user-attachments/assets/53e9b0ea-b33e-4034-9b1f-23f913222d81" />

**Comment by neilmehta24** [comment]
Thanks for bringing this to our attention. The [MLX core library](https://github.com/ml-explore/mlx) does indeed support a top-k matrix operation, but the [MLX LLM library](https://github.com/ml-explore/mlx-examples/tree/main/llms) does not support top-k sampling. Here are the supported generation/sampling options as of today https://github.com/ml-explore/mlx-examples/blob/dfa4dd6/llms/mlx_lm/utils.py#L200-L215 . Please track this issue https://github.com/ml-explore/mlx-examples/issues/1167 for adding support in `mlx_lm`

**Comment by mikemykhaylov** [comment]
Looks like the upstream PR got merged, any chance we could have the sample in the MLX engine?

**Comment by neilmehta24** [comment]
@mikemykhaylov FYI this is slightly delayed since we need to update the LM Studio UI to inform users about the limitation described here https://github.com/ml-explore/mlx-examples/issues/1219

**Comment by mikemykhaylov** [comment]
@neilmehta24 Thank you for writing the fix for the linked issue! Any chance we could be expecting this in LM Studio?

**Comment by neilmehta24** [comment]
@mikemykhaylov yes, this is unblocked now. I just have to figure out how to prevent another reversion here https://github.com/lmstudio-ai/lmstudio-js/pull/192 . This isn't the top priority for us, but you can still expect to see it soon

**Comment by neilmehta24** [comment]
Completed. This feature will be available in the LM Studio 0.3.15+5 or greater
